dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/cache/TimeToLive.java

```diff
@@ -0,0 +1,170 @@
+ *
+ * Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ * Neither the name of the HISP project nor the names of its contributors may
+ * be used to endorse or promote products derived from this software without
+ * specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package org.hisp.dhis.analytics.cache;
+
+import static java.time.LocalDateTime.now;
+import static java.time.LocalDateTime.ofInstant;
+import static java.time.ZoneOffset.UTC;
+import static java.time.temporal.ChronoUnit.DAYS;
+import static java.time.temporal.ChronoUnit.YEARS;
+import static java.util.concurrent.TimeUnit.MINUTES;
+import static org.hisp.dhis.analytics.cache.TimeToLive.IntervalOfMilliseconds.FIFTEEN_MINUTES;
+import static org.hisp.dhis.analytics.cache.TimeToLive.IntervalOfMilliseconds.FIVE_MINUTES;
+import static org.hisp.dhis.analytics.cache.TimeToLive.IntervalOfMilliseconds.FORTY_MINUTES;
+import static org.hisp.dhis.analytics.cache.TimeToLive.IntervalOfMilliseconds.SIXTY_MINUTES;
+import static org.hisp.dhis.analytics.cache.TimeToLive.IntervalOfMilliseconds.TWO_MINUTES;
+import static org.hisp.dhis.analytics.cache.TimeToLive.Periods.MONTHLY;
+import static org.hisp.dhis.analytics.cache.TimeToLive.Periods.QUARTERLY;
+import static org.hisp.dhis.analytics.cache.TimeToLive.Periods.SIX_MONTHS;
+import static org.hisp.dhis.analytics.cache.TimeToLive.Periods.WEEKLY;
+import static org.hisp.dhis.analytics.cache.TimeToLive.Periods.YEARLY;
+import static org.springframework.util.Assert.notNull;
+
+import java.time.Instant;
+import java.util.Map;
+
+import org.hisp.dhis.analytics.DataQueryParams;
+
+import com.google.common.collect.ImmutableMap;
+
+public class TimeToLive
+    implements
+    Computable
+{
+
+    private final DataQueryParams params;
+
+    static final Map<Periods, IntervalOfMilliseconds> EXPIRATION_TIME_TABLE = ImmutableMap.of(
+        WEEKLY, TWO_MINUTES,
+        MONTHLY, FIVE_MINUTES,
+        QUARTERLY, FIFTEEN_MINUTES,
+        SIX_MONTHS, FORTY_MINUTES,
+        YEARLY, SIXTY_MINUTES );
```

Comment:
I think we need a configurable solution. E.g. currently the general analytics cache is configurable in seconds. There should be some sensible/simple way of configuring the different cache TTL for various time thresholds. Let's discuss.

Reply:
Configuration settings added to SettingKey and managed through the DefaultSystemSettingManager.

TTL are now defined in seconds.

Reply:
Not used anymore, removed